In [1]:
!pip install numpy

In [2]:
import numpy as np

### Creating a Vector
A vector has a both Magnitude and Direction.

In [5]:
vector_row = np.array([1, 2, 3])
vector_row

array([1, 2, 3])

In [11]:
vector_col = np.array([
    [1, 2, 3],
    [4, 5, 6],
    [6, 7, 8]
])
print(vector_col)

[[1 2 3]
 [4 5 6]
 [6 7 8]]


### Creating a Matrix

In [18]:
matrix = np.array([
    [1,2],
    [1,2],
    [1,2]
])
matrix

array([[1, 2],
       [1, 2],
       [1, 2]])

In [20]:
# Numpy actually has a dedicated matrix data structure
matrix_obj = np.mat([
    [1,2],
    [1,2],
    [1,2]
])
matrix_obj

matrix([[1, 2],
        [1, 2],
        [1, 2]])

#### Matrix data structure is not recommended for 2 reasons:
<div style="font-size:15px">
Arrays are the de facto standard data structure of numpy. <br>
Majority of numpy operations returns array, not matrix objects.
</div>

### Creating a Sparse Matrix
<div style="font-size:15px">
    A frequent situation in machine learning is having huge amount of data where most of the datas are zeros. <br> Sparse matrices only store non-zero elements and assumes all the other values will be 0, leading to significant computational savings.
</div>

In [54]:
from scipy import sparse
matrix = np.array([
    [33, 1, 0],
    [0, 0, 10],
    [3, 0, 0]
])
matrix_sparse = sparse.csr_matrix(matrix) # Compressed sparse row matrices where non-zero elements and their positions are stored.
print(matrix_sparse)

  (0, 0)	33
  (0, 1)	1
  (1, 2)	10
  (2, 0)	3


In [72]:
matrix_large = np.array([
    [33, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 10],
    [3, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
])
matrix_large_sparse = sparse.csr_matrix(matrix_large)
print(matrix_large_sparse) # Row-wise representation

  (0, 0)	33
  (0, 1)	1
  (1, 19)	10
  (2, 0)	3


<div style="font-size:15px">
    Despite the fact that we have added so many zeros in the matrix, the sparse matrix representation remains the same. That is, addition of zero elements did not change the size of the sparse matrix.
</div>

In [75]:
matrix_large_sparse = sparse.csc_matrix(matrix_large)
print(matrix_large_sparse) # Column-wis representation

  (0, 0)	33
  (2, 0)	3
  (0, 1)	1
  (1, 19)	10


In [77]:
print(matrix_large_sparse.toarray()) # Converting the sparse matrix back to original 2-D array

[[33  1  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0]
 [ 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0 10]
 [ 3  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0]]


In [79]:
print(f"Size in bytes of original matrix: {matrix_large.nbytes}")
print(f"Size in bytes of compressed sparse column matrix: {matrix_large_sparse.data.nbytes + matrix_large_sparse.indptr.nbytes + matrix_large_sparse.indices.nbytes}")

Size in bytes of original matrix: 240
Size in bytes of compressed sparse column matrix: 116


<div style="font-size:15px">
    The standard numpy representation takes 240 bytes while the sparse representation takes the half of actual size i.e., 116 bytes.
</div>

### Pre-allocating the Numpy Arrays
<div style="font-size:15px">
    Numpy has the functions for generating vectors and matrices of any size using 0s, 1s or values of your choice.
</div>

In [98]:
vector = np.zeros(shape = 5)
print(vector, vector.dtype, sep="\n")

[0. 0. 0. 0. 0.]
float64


In [108]:
matrix = np.full(shape = (3,3), fill_value = 45, dtype="float")
print(matrix, matrix.dtype, sep="\n")

[[45. 45. 45.]
 [45. 45. 45.]
 [45. 45. 45.]]
float64


<div style="font-size:15px">
    Generating the array pre-filled with data is useful for number of purposes, such as making code more performant or having sythetic data to test the algorithms with.
</div>

### Selecting Elements
<div style="font-size:15px">
    NumPy array makes it easy to select the elements from vector/array.
</div>

In [8]:
vector = np.array([1, 2, 3, 4, 5, 6])
matrix = np.array([
    [1, 2, 3],
    [4, 5, 6],
    [7, 8, 9]
])

In [10]:
vector[2]

3

In [14]:
matrix[1,1]

5

In [16]:
vector[:]

array([1, 2, 3, 4, 5, 6])

In [18]:
vector[3:]

array([4, 5, 6])

In [20]:
vector[-1]

6

In [22]:
vector[::-1]

array([6, 5, 4, 3, 2, 1])

In [24]:
matrix[:2,:]

array([[1, 2, 3],
       [4, 5, 6]])

In [32]:
matrix[:,1]

array([2, 5, 8])

In [28]:
matrix[:,1:2]

array([[2],
       [5],
       [8]])

### Describing the Matrix
<div style="font-size:15px">
    Using the shape, size and ndim attributes of numpy object
</div>

In [36]:
matrix.shape

(3, 3)

In [38]:
matrix.size

9

In [40]:
matrix.ndim

2

### Apply Functions over Each Element of the Numpy Array
<div style="font-size:15px">
    NumPy's Vectorize method
</div>

In [46]:
vectorize_method = np.vectorize(lambda x : x + 100)
vectorize_method

In [48]:
vectorize_method(matrix)

array([[101, 102, 103],
       [104, 105, 106],
       [107, 108, 109]])

In [50]:
# Using Broadcasting
matrix + 100

array([[101, 102, 103],
       [104, 105, 106],
       [107, 108, 109]])

### Finding the maximum and minimum values

In [54]:
np.max(matrix)

9

In [58]:
np.min(matrix)

1

In [62]:
# Find the maximum element in each column.
np.max(matrix, axis=0)

array([7, 8, 9])

In [64]:
# Find the maximum element in each row.
np.max(matrix, axis=1)

array([3, 6, 9])

### Calculating the Average, Variance & Standard Deviation

In [68]:
np.mean(matrix)

5.0

In [70]:
np.var(matrix)

6.666666666666667

In [72]:
np.std(matrix)

2.581988897471611

### Reshaping the Array

In [75]:
matrix = np.array([
    [1, 2, 3, 4],
    [5, 6, 7, 8],
    [10, 11, 12, 13]
])
print(matrix)

[[ 1  2  3  4]
 [ 5  6  7  8]
 [10 11 12 13]]


In [77]:
matrix.reshape(4,3)

array([[ 1,  2,  3],
       [ 4,  5,  6],
       [ 7,  8, 10],
       [11, 12, 13]])

In [79]:
matrix.reshape(6,2)

array([[ 1,  2],
       [ 3,  4],
       [ 5,  6],
       [ 7,  8],
       [10, 11],
       [12, 13]])

In [83]:
# matrix.reshape(3,5)
# ValueError                                Traceback (most recent call last)
# Cell In[81], line 1
# ----> 1 matrix.reshape(3,5)

# ValueError: cannot reshape array of size 12 into shape (3,5)

<div style="font-size:15px">
    One useful argument in reshape is -1, which effetively means "As Many As Needed", so reshape (1, -1) means one row and as many columns as needed.
</div>

In [88]:
matrix.reshape(1, -1)

array([[ 1,  2,  3,  4,  5,  6,  7,  8, 10, 11, 12, 13]])

In [90]:
matrix.reshape(-1, 1)

array([[ 1],
       [ 2],
       [ 3],
       [ 4],
       [ 5],
       [ 6],
       [ 7],
       [ 8],
       [10],
       [11],
       [12],
       [13]])

In [99]:
matrix.reshape(12)

array([ 1,  2,  3,  4,  5,  6,  7,  8, 10, 11, 12, 13])

### Transposing a Vector/Matrix

In [102]:
print(matrix)

[[ 1  2  3  4]
 [ 5  6  7  8]
 [10 11 12 13]]


In [104]:
print(matrix.T)

[[ 1  5 10]
 [ 2  6 11]
 [ 3  7 12]
 [ 4  8 13]]


### Flattening a Matrix

In [108]:
matrix.flatten()

array([ 1,  2,  3,  4,  5,  6,  7,  8, 10, 11, 12, 13])

<div style="font-size:16px">
    One more method to flatten the array is ravel method. Unlike flatten method, that returns the copy of the orignal object. ravel() operates on the same object and is slightly more faster than flatten(). It also flattens the list of arrays, that is not possible to do using flatten() method.
</div>

In [112]:
a = np.array([
    [1, 2],
    [3, 4]
])
b = np.array([
    [5, 6],
    [7, 8]
])

c = [a, b]
print(c)

[array([[1, 2],
       [3, 4]]), array([[5, 6],
       [7, 8]])]


In [116]:
np.ravel(c)

array([1, 2, 3, 4, 5, 6, 7, 8])

### Finding the rank of the matrix.

In [119]:
np.linalg.matrix_rank(matrix)

2

### Getting the diagonal of the matrix

In [126]:
print(matrix)

[[ 1  2  3  4]
 [ 5  6  7  8]
 [10 11 12 13]]


In [122]:
matrix.diagonal()

array([ 1,  6, 12])

In [130]:
# Return the diagonal one above the main diagonal
matrix.diagonal(offset=1)

array([ 2,  7, 13])

In [134]:
# Return the diagonal one below the main diagonal
matrix.diagonal(offset=-1)

array([ 5, 11])

### Getting the trace of the matrix

In [138]:
matrix.trace()

19

In [140]:
np.sum(matrix.diagonal())

19

In [146]:
a = np.array([1,2,3,4])
b = np.array([5,6,7,8])
np.dot(a, b)

70

In [148]:
a @ b

70

### Adding Subtracting & Multiplying the matrices

In [151]:
mat_a = np.array([[1,2,3],[4,5,6],[7,8,9]])
mat_b = np.array([[1,1,1],[9,9,9],[3,3,3]])

In [153]:
np.add(mat_a, mat_b)

array([[ 2,  3,  4],
       [13, 14, 15],
       [10, 11, 12]])

In [155]:
np.subtract(mat_a, mat_b)

array([[ 0,  1,  2],
       [-5, -4, -3],
       [ 4,  5,  6]])

In [157]:
mat_a + mat_b

array([[ 2,  3,  4],
       [13, 14, 15],
       [10, 11, 12]])

In [160]:
mat_a @ mat_b

array([[ 28,  28,  28],
       [ 67,  67,  67],
       [106, 106, 106]])

In [162]:
np.dot(mat_a, mat_b)

array([[ 28,  28,  28],
       [ 67,  67,  67],
       [106, 106, 106]])

In [164]:
mat_a*mat_b

array([[ 1,  2,  3],
       [36, 45, 54],
       [21, 24, 27]])

### Inverting the Matrix

In [171]:
matrix = np.array([
    [1, 2, 3],
    [4, 5, 6],
    [7, 8, 9]
])
np.linalg.inv(matrix)

array([[ 3.15251974e+15, -6.30503948e+15,  3.15251974e+15],
       [-6.30503948e+15,  1.26100790e+16, -6.30503948e+15],
       [ 3.15251974e+15, -6.30503948e+15,  3.15251974e+15]])

In [177]:
matrix @ np.linalg.inv(matrix)

array([[ 0. ,  1. , -0.5],
       [ 0. ,  2. , -1. ],
       [ 0. ,  3. ,  2.5]])

### Generating Random values

In [180]:
np.random.seed(12345) # To reproduce the same random values

In [184]:
# Generate 3 random floats from 0 and 1
np.random.random(3)

array([0.20456028, 0.56772503, 0.5955447 ])

In [188]:
# Generate 3 random integers between 0 and 10
np.random.randint(0, 11, 3)

array([7, 6, 0])

In [196]:
# Draw three numbers from a normal distribution with mean 0.0 and standard deviation of 1.0
np.random.normal(0, 1, 3)

array([ 1.18031326,  1.41795968, -1.00970237])

In [192]:
# Draw three numbers from a logistic distribution with mean 0.0 and scale of 1.0
np.random.logistic(0, 1, 3)

array([2.28855708, 0.22655328, 1.6576274 ])

In [194]:
# Draw 3 numbers greater than or equal to 1 and less than 2
np.random.uniform(1, 2, 3)

array([1.05048796, 1.80623485, 1.93081569])

## Done with Day 1 :)